soc/xtensa/intel_adsp/ace_v1x/multiprocessing.c

```diff
@@ -0,0 +1,205 @@
+		/* DINT has one bit per IPC, unmask only IPC "Ax" on core "x" */
+		MTL_DINT[i].ie[MTL_INTL_IDCA] = BIT(i);
+
+		/* Agent A should signal only BUSY interrupts */
+		MTL_P2P_IPC[i].agents[0].ipc.ctl = BIT(0); /* IPCTBIE */
+	}
+
+	/* Set the core 0 active */
+	soc_cpus_active[0] = true;
+}
+
+void soc_start_core(int cpu_num)
+{
+	int retry = CORE_POWER_CHECK_NUM;
+	if (cpu_num > 0) {
+		/* Initialize the ROM jump address */
+		uint32_t* rom_jump_vector = (uint32_t *) ROM_JUMP_ADDR;
+		*rom_jump_vector = (uint32_t) z_soc_mp_asm_entry;
+		z_xtensa_cache_flush(rom_jump_vector, sizeof(*rom_jump_vector));
+
+		/* Tell the ACE ROM that it should use secondary core flow */
+		MTL_PWRBOOT.bootctl[cpu_num].battr |= MTL_PWRBOOT_BATTR_LPSCTL_BATTR_SLAVE_CORE;
+	}
+
+	MTL_PWRBOOT.capctl[cpu_num].ctl |= MTL_PWRBOOT_CTL_SPA;
+
+	/* Waiting for power up */
+	while (~(MTL_PWRBOOT.capctl[cpu_num].ctl & MTL_PWRBOOT_CTL_CPA) && --retry) {
+		 k_busy_wait(CORE_POWER_CHECK_DELAY);
+	}
+
+	if(!retry) {
+		__ASSERT(false, "soc_start_core() secondary core has not powered up");
+	}
+}
+
+void soc_mp_startup(uint32_t cpu)
+{
+	/* Must have this enabled always */
+	z_xtensa_irq_enable(ACE_INTC_IRQ);
+
+	/* Prevent idle from powering us off */
+	MTL_PWRBOOT.bootctl[cpu].bctl |=
+		MTL_PWRBOOT_BCTL_WAITIPCG | MTL_PWRBOOT_BCTL_WAITIPPG;
+}
+
+void arch_sched_ipi(void)
+{
+	uint32_t curr = processor_id_reg();
+
+	/* Signal agent B[n] to cause an interrupt from agent A[n] */
+	for (int core = 0; core < CONFIG_MP_NUM_CPUS; core++) {
+		if (core != curr && soc_cpus_active[core]) {
+			MTL_P2P_IPC[core].agents[1].ipc.idr = CAVS_IPC_BUSY;
+		}
+	}
+}
+
+int soc_adsp_halt_cpu(int id)
+{
+	int retry = CORE_POWER_CHECK_NUM;
+
+	CHECKIF(processor_id_reg() != 0) {
+		return -EINVAL;
+	}
+
+	CHECKIF(id <= 0 || id >= CONFIG_MP_NUM_CPUS) {
+		return -EINVAL;
+	}
+
+	CHECKIF(soc_cpus_active[id]) {
+		return -EINVAL;
+	}
+
+	MTL_PWRBOOT.capctl[id].ctl &= ~MTL_PWRBOOT_CTL_SPA;
+
+	/* Waiting for power off */
+	while (MTL_PWRBOOT.capctl[id].ctl & MTL_PWRBOOT_CTL_CPA && --retry)
+		 k_busy_wait(CORE_POWER_CHECK_DELAY);
+
+	if(!retry) {
+		__ASSERT(false, "soc_adsp_halt_cpu() secondary core has not powered down");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+#if defined(CONFIG_PM)
+/* Invoke Low Power/System Off specific Tasks */
+__weak void pm_state_set(enum pm_state state, uint8_t substate_id)
+{
+	ARG_UNUSED(substate_id);
+	uint32_t cpu = processor_id_reg();
+
+	switch (state) {
+	case PM_STATE_SOFT_OFF:/* D3 */
+		MTL_PWRBOOT.bootctl[cpu].bctl &= ~MTL_PWRBOOT_BCTL_WAITIPCG;
+		soc_cpus_active[cpu] = false;
+		z_xtensa_cache_flush_inv_all();
+		if (cpu == 0) {
+			/* FIXME: this value should come from MM */
+			uint32_t hpsram_mask[1] = { 0x3FFFFF };
+			ace_power_down(true, uncache_to_cache(&hpsram_mask[0]), true);
+		} else {
+			k_cpu_idle();
+		}
+
+		break;
+	case PM_STATE_SUSPEND_TO_IDLE: /* D0ix */
+		__fallthrough;
+	case PM_STATE_RUNTIME_IDLE:/* D0 */
+		k_cpu_idle();
+		break;
```

Comment:
we can go even further, if the target just implement `PM_STATE_SOFT_OFF` it does not need to declare other states in DT and handle them here. The idle thread will automatically call `k_cpu_idle()`. My understanding is that there is a desire to map all states that the target supports even if the implementation is not really doing anything special. It is more complicated than it needs to be, but it is not a blocker for me.

Reply:
@andyross @ceolin I don't understand why it's a problem now, since these states have to be implemented anyway. Of course you can do it as you suggested and add all the code for each state separately.

Reply:
It is not a problem for me. I just said that this can be simpler.